src/libs/sfdk/dockervirtualmachine.cpp

```diff
@@ -0,0 +1,494 @@
+        const QObject *context, const Functor<bool> &functor)
+{
+    Q_UNUSED(ruleName)
+    Q_UNUSED(context)
+    Q_UNUSED(functor)
+    QTC_CHECK(false);
+}
+
+void DockerVirtualMachinePrivate::doSetReservedPortForwarding(VirtualMachinePrivate::ReservedPort which,
+        quint16 port, const QObject *context, const Functor<bool> &functor)
+{
+    QPair<quint16, quint16>& mapping = m_reservedPortMappings[which];
+    mapping.first = port;
+
+    buildWithLabel(label(which), QString::number(port), context, functor);
+}
+
+void DockerVirtualMachinePrivate::doSetReservedPortListForwarding(
+        VirtualMachinePrivate::ReservedPortList which, const QList<Utils::Port> &ports,
+        const QObject *context, const Functor<const QMap<QString, quint16> &, bool> &functor)
+{
+    Q_UNUSED(which)
+    Q_UNUSED(ports)
+    Q_UNUSED(context)
+    Q_UNUSED(functor)
+    QTC_CHECK(false);
+}
+
+void DockerVirtualMachinePrivate::doRestoreSnapshot(const QString &snapshotName, const QObject *context,
+        const Functor<bool> &functor)
+{
+    Q_UNUSED(snapshotName)
+    Q_UNUSED(context)
+    Q_UNUSED(functor)
+    QTC_CHECK(false);
+}
+
+QString DockerVirtualMachinePrivate::label(VirtualMachinePrivate::SharedPath which)
+{
+    QString label;
+    switch (which) {
+    case SharedSrc:
+        label = Constants::BUILD_ENGINE_SHARED_SRC;
+        break;
+    case SharedHome:
+        label = Constants::BUILD_ENGINE_SHARED_HOME;
+        break;
+    case SharedConfig:
+        label = Constants::BUILD_ENGINE_SHARED_CONFIG;
+        break;
+    case SharedSsh:
+        label = Constants::BUILD_ENGINE_SHARED_SSH;
+        break;
+    case SharedTargets:
+        label = Constants::BUILD_ENGINE_SHARED_TARGET;
+    break;
+    }
+    return label;
+}
+
+QString DockerVirtualMachinePrivate::label(VirtualMachinePrivate::ReservedPort which)
+{
+    QString label;
+    switch (which) {
+    case SshPort:
+        label = Constants::BUILD_ENGINE_SSH_PORT;
+        break;
+    case WwwPort:
+        label = Constants::BUILD_ENGINE_WWW_PORT;
+        break;
+    }
+    return label;
+}
+
+void DockerVirtualMachinePrivate::buildWithLabel(const QString& key, const QString& value,
+        const QObject *context, const Functor<bool> &functor)
+{
+    Q_Q(DockerVirtualMachine);
+    QStringList arguments;
+    arguments.append(COMMAND_BUILD);
+    arguments.append(DOCKERFILE_PREFIX + Sdk::installationPath() + DOCKERFILE_POSTFIX);
+    arguments.append(LABEL_PREFIX + key + "=" + value);
+    arguments.append(TAG);
+    arguments.append(q->name());
+    arguments.append(Sdk::installationPath());
+
+    const QPointer<const QObject> context_{context};
+
+    auto runner = std::make_unique<ProcessRunner>(DockerManager::dockerPath(), arguments);
+    QObject::connect(runner.get(), &ProcessRunner::done, q, [context_, functor, process = runner->process()](bool ok) {
+        callIf(context_, functor, ok);
+    });
```

Comment:
`QObject::connect(runner.get(), &ProcessRunner::done, context, functor);` should be enough here.

Reply:
fixed